### PR TITLE
Paint undo

### LINKED ISSUE #290: MFPaint: Undo (and Redo?)
It would be nice to be able to undo and redo recent edits

Bonus points if I can undo and redo palette changes!

### PR DESCRIPTION
**Pull request by dylwhich**
Uses SPIRAM to allow up to 984 actions to be undone!
I renamed the linked list implementation used by the emu (`emu/src/list/list.h`) from `list_t` to `linked_list_t` because both that and the `list_t` from `main/utils/linked_list.h` ended up included in the main and caused conflicts.

Resolves #290

### CODE REVIEW

**Comment by AEFeinstein** [comment]
Without even reading the changes, I do not like multiple linked list implementations in the same source. Any reason not to remove the implementation in `emu` entirely? I probably copied it from the IDF to get things off the ground...

**Review by AEFeinstein — APPROVED** [review]
Other than the list thing, seems fine. A few functions got returns without updated doxygen, if you care about that sort of thing.

**Review by AEFeinstein — CHANGES_REQUESTED** [review]
Other than the list thing, seems fine. A few functions got returns without updated doxygen, if you care about that sort of thing.

**Comment by dylwhich** [comment]
> Without even reading the changes, I do not like multiple linked list implementations in the same source. Any reason not to remove the implementation in `emu` entirely? I probably copied it from the IDF to get things off the ground...

Ah sure, I can do that. I just didn't want to swap it out without knowing the rationale.



> Other than the list thing, seems fine. A few functions got returns without updated doxygen, if you care about that sort of thing.

Oops, I meant to do that and forgot. Will fix

File: emu/src/emu_sensors.c [hunk 1]
```diff
@@ -65,23 +66,23 @@ void deinitButtons(void)
...
 
-	while (NULL != (node = list_lpop(touchQueue)))
+	while (NULL != (val = shift(touchQueue)))
 	{
 		// Free everything
-		free(node->val);
-		free(node);
+		free(val);
 	}
+	clear(buttonQueue);
```
Comment (AEFeinstein) [inline emu/src/emu_sensors.c:85]: Copy paste bug

**Review by AEFeinstein — CHANGES_REQUESTED** [review]
Maybe do some valgrind or asan to exercise the linked list and paint changes too?

**Review by AEFeinstein — APPROVED** [review]

**Pull request merged**